app/services/charges/charge_models/prorated_graduated_service.rb

```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
+module Charges
+  module ChargeModels
+    class ProratedGraduatedService < Charges::ChargeModels::BaseService
+      protected
+
+      def ranges
+        properties['graduated_ranges']&.map(&:with_indifferent_access)
+      end
+
+      def compute_amount
+        full_units = per_event_aggregation_result.event_aggregation
+        prorated_units = per_event_aggregation_result.event_prorated_aggregation
+
+        index = 0
+        full_sum = 0
+        overflow = 0
+        ranges.reduce(0) do |result_amount, range|
+          prorated_sum = 0
+          flat_amount = BigDecimal(range[:flat_amount])
+          per_unit_amount = BigDecimal(range[:per_unit_amount])
+
+          # NOTE: Add flat amount to the total
+          result_amount += flat_amount unless units.zero?
+
+          # Calculate prorated value inside the range. Which events are taken into account
+          # for certain range depends on comparing full number of units and range boundaries
+          loop do
```

Comment:
:scream: I'm a bit uncomfortable with this operator. Do you think there is a way to get rid of it? I think it's better to have an explicit output condition

Reply:
Refactored a bit, hope so that it's more readable now